### PR TITLE
PWN-3114 - Device token send retry

### PR DESCRIPTION
**Pull request by kuznetsov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-3114

## Description of Work

Retry device token send request after a minute in case of ServerException

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt [hunk 1]
```diff
@@ -3,22 +3,28 @@ package org.p2p.wallet.push_notifications.ineractor
...
 class PushNotificationsInteractor(
     private val deviceTokenRepository: DeviceTokenRepository,
     private val pushTokenRepository: PushTokenRepository,
     private val tokenKeyProvider: TokenKeyProvider,
     private val sharedPreferences: SharedPreferences,
+    private val appScope: AppScope
 ) {
 
-    suspend fun updateDeviceToken(): DeviceToken {
+    suspend fun updateDeviceToken(retries: Int = 1) {
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:27]: i'd prefer to move 1 to constant and use/reuse it

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt (hunk 1 above)
Comment (kuznetsov) [reply app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:27]: It's a one use value, so it's no need for constant

Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:27]: Agree with @kabramovp2p, move to variable or to constant

File: app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt [hunk 2]
```diff
@@ -35,7 +41,20 @@ class PushNotificationsInteractor(
...
+            retryTokenSend(retries)
+        }
+    }
+
+    private fun retryTokenSend(retries: Int) {
+        if (retries < 1) return
+
+        appScope.launch {
+            delay(TOKEN_SEND_RETRY_DELAY)
+            updateDeviceToken(retries - 1)
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:56]: maybe we move retries - 1 to separate line ?

File: app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt [hunk 3]
```diff
@@ -3,22 +3,28 @@ package org.p2p.wallet.push_notifications.ineractor
...
+import org.p2p.wallet.common.di.AppScope
+import org.p2p.wallet.infrastructure.network.data.ServerException
 import org.p2p.wallet.infrastructure.network.provider.TokenKeyProvider
 import org.p2p.wallet.push_notifications.model.DeviceInfo
 import org.p2p.wallet.push_notifications.model.DeviceToken
 import org.p2p.wallet.push_notifications.repository.DeviceTokenRepository
 import org.p2p.wallet.push_notifications.repository.PushTokenRepository
 
 private const val KEY_DEVICE_TOKEN = "KEY_DEVICE_TOKEN"
+private const val TOKEN_SEND_RETRY_DELAY = 60000L // ms
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:17]: you can put MS in the const name or rewrite using TimeUnit

File: app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt (hunk 2 above)
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:56]: you are not retrying sending token here, but rewriting shared prefs too

File: app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt [hunk 4]
```diff
@@ -35,7 +41,20 @@ class PushNotificationsInteractor(
...
+    }
+
+    private fun retryTokenSend(retries: Int) {
+        if (retries < 1) return
+
+        appScope.launch {
+            delay(TOKEN_SEND_RETRY_DELAY)
+            updateDeviceToken(retries - 1)
+        }
     }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:58]:
```suggestion
        appScope.launch {
            sendTokenToServer(deviceToken, retries)
        }
    }

    private suspend fun sendTokenToServer(deviceToken: DeviceToken, retries: Int) {
        val noRetryNeeded = retries == 0
        if (noRetryNeeded) return

        try {
            deviceTokenRepository.sendDeviceToken(deviceToken)
        } catch (e: ServerException) {
            delay(TOKEN_SEND_RETRY_DELAY)
            sendTokenToServer(deviceToken, retries - 1)
        }
    }
```

Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:58]: your first send try was not in appScope, so I moved it
i extracted sending method so it's recursion in only one place, plus no shared preferences logic included any longer

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt [hunk 5]
```diff
@@ -35,7 +41,20 @@ class PushNotificationsInteractor(
             clientId = tokenKeyProvider.publicKey
         )
 
-        return deviceTokenRepository.sendDeviceToken(deviceToken)
+        try {
+            deviceTokenRepository.sendDeviceToken(deviceToken)
+        } catch (e: ServerException) {
+            retryTokenSend(retries)
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:47]: Please, add timber logs

File: app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt (hunk 1 above)
Comment (kuznetsov) [reply app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:27]: Okay then

File: app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt (hunk 4 above)
Comment (kuznetsov) [reply app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:58]: Thank, that'a a good point!

**Review by 1sakov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt [hunk 6]
```diff
@@ -3,22 +3,29 @@ package org.p2p.wallet.push_notifications.ineractor
...
 import org.p2p.wallet.infrastructure.network.provider.TokenKeyProvider
 import org.p2p.wallet.push_notifications.model.DeviceInfo
 import org.p2p.wallet.push_notifications.model.DeviceToken
 import org.p2p.wallet.push_notifications.repository.DeviceTokenRepository
 import org.p2p.wallet.push_notifications.repository.PushTokenRepository
+import timber.log.Timber
 
 private const val KEY_DEVICE_TOKEN = "KEY_DEVICE_TOKEN"
+private const val TOKEN_SEND_RETRY_DELAY_MS = 60000L
+private const val RETRIES_NUMBER = 1
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:17]:
```suggestion
private const val DEFAULT_RETRIES_NUMBER = 1
```

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt [hunk 7]
```diff
@@ -35,7 +42,13 @@ class PushNotificationsInteractor(
             clientId = tokenKeyProvider.publicKey
         )
 
-        return deviceTokenRepository.sendDeviceToken(deviceToken)
+        try {
+            deviceTokenRepository.sendDeviceToken(deviceToken)
+        } catch (e: ServerException) {
+            Timber.e(e, "Error sending device token $token to server")
+            delay(TOKEN_SEND_RETRY_DELAY_MS)
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:49]: It's not a good idea to block thread resources for one minute, i think it's insufficient way and we need to go with another way of implementation

Comment (kuznetsov) [reply app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt:49]: Let's not overhead the solution for now and check how it performs on the QA stage

**Review by davran1001 — APPROVED** [review]

**Pull request merged**